### PR TITLE
Force Shield (+Mass version) spell

### PR DESCRIPTION
**Pull request by Kert**
"Creates a Kinetic shield over the target ally"
The shield has as much hp as one creature from the stack affected by the spell.
The shield absorbs damage until it's hp becomes 0, the remaining damage goes as usual.
You can't stack force shields.
There is also a Mass Force Shield spell.

### CODE REVIEW

**Comment by Kert** [comment]
Do not review yet needs one more fix

**Comment by jkoppel** [comment]
Is there an easy way to show the force field HP remaining to the user?

File: src/cpp/shared/combat/combat_spells.cpp [hunk 1]
```diff
@@ -635,6 +638,14 @@ void combatManager::CastSpell(int proto_spell, int hexIdx, int isCreatureAbility
     case SPELL_PLASMA_CONE:
       this->Fireball(hexIdx, SPELL_PLASMA_CONE);
       break;
+    case SPELL_FORCE_SHIELD:
+      this->ShowSpellMessage(isCreatureAbility, proto_spell, stack);
+
+      stack->DispelGood();
```
[the remark this answers is not in the document]
Comment (Kert) [reply src/cpp/shared/combat/combat_spells.cpp]: > Remove buffs before casting the spell
From the commit name
If this shouldn't be done just tell me

File: src/cpp/shared/combat/army.cpp [hunk 2]
```diff
@@ -1868,7 +1873,7 @@ void army::PowEffect(int animIdx, int a3, int a4, int a5) {
                 if (creature->animationType != ANIMATION_TYPE_STANDING && creature->animationType != ANIMATION_TYPE_DYING) {
                   creature->animationType = ANIMATION_TYPE_STANDING;
                   creature->animationFrame = 0;
-                  creature->effectStrengths[15] = 1;
+                  creature->field_3 = -1;
```
[the remark this answers is not in the document]
Comment (Kert) [reply src/cpp/shared/combat/army.cpp]: That's from one of the previous spell branches I think
Original code had this effectStrengths field change code in this function. Even though this function's only purpose is to show some visual effects on attack/spell done on creatures
My guess is the decompilation was a bit wrong here
In any case as far as I remember I tracked down which variable I can use instead and this was my solution for the bug https://github.com/jkoppel/project-ironfist/commit/ecfcd64fe536b4e0855f0f435eb8c04b2fab802d

File: src/cpp/shared/combat/combat_spells.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkoppel) [reply src/cpp/shared/combat/combat_spells.cpp]: What's the reason for doing this?

**Comment by Kert** [comment]
> Is there an easy way to show the force field HP remaining to the user?
Should be not very hard to make a second box similar to one that shows the quantity of creatures above or below also with a "positive" color saying something like "+100"

File: src/cpp/shared/combat/combat_spells.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (Kert) [reply src/cpp/shared/combat/combat_spells.cpp]: I have no idea. Maybe we've discussed it previously and you asked to add this?

**Comment by Kert** [comment]
shieldHP
How do you like this?
I used spell effect small icon with red palette substitution to make white font stand out
I can also add shield hp numbers to info window and creature info screen if you like this design

File: src/cpp/shared/combat/army.cpp [hunk 3]
```diff
@@ -2456,3 +2478,188 @@ int army::GetStraightLineDistanceToHex(int hex) {
...
+    } else {
+      inRedrawZone = gpCombatManager->combatScreenIcons[1]->CombatClipDrawToBuffer(stackNumXOffset, offsetY, 10, &this->stackSizeDispBounds, 0, 0, 0, 0);
+    }
+    if(gIronfistExtra.combat.stack.forceShieldHP[this] > 0) {
+      const int FORCE_SHIELD_ICON_X_OFFSET = 2;
+      const int FORCE_SHIELD_ICON_Y_OFFSET = -28;
+      const int FORCE_SHIELD_HP_BOX_Y_OFFSET = -24;
+
+      icon* shieldIcon = gpResourceManager->GetIcon("SPELLINF.ICN");
+      shieldIcon->CombatClipDrawToBuffer(stackNumXOffset + FORCE_SHIELD_ICON_X_OFFSET, offsetY + FORCE_SHIELD_ICON_Y_OFFSET, 10, &this->stackSizeDispBounds, 0, 0, gColorTableRed, 0);
```
Comment (jkoppel) [inline src/cpp/shared/combat/army.cpp]: Does this work when the player is the defender?

File: src/cpp/shared/combat/combat_spells.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (jkoppel) [reply src/cpp/shared/combat/combat_spells.cpp]: Remove this line.

File: src/cpp/shared/spell/spells.cpp [hunk 4]
```diff
@@ -79,7 +79,9 @@ char *gSpellDesc[] =
   "{Awareness}\n\nExplores a large area around your hero.",
   "{Shadow Mark}\n\nMarked creature receives 150% damage from your troops for 1 turn",
   "{Marksman Pierce}\n\nTarget creature receives 1000 damage and is dazed for 1 turn",
-  "{Plasma Cone}\n\nCreates a cone of hot plasma in the direction of the target"
+  "{Plasma Cone}\n\nCreates a cone of hot plasma in the direction of the target",
+  "{Force Shield}\n\nCreates a Kinetic shield over the target ally",
```
Comment (jkoppel) [inline src/cpp/shared/spell/spells.cpp]: Either all lowercase "kinetic shield" or all uppercase "Kinetic Shield."

**Comment by jkoppel** [comment]
I like that display a lot, although it's unfortunate that it blocks some creature graphics.

**Comment by jkoppel** [comment]
In some of my early tests, I saw the mass spell not being cast on dwarves. However, in later testing, I was unable to reproduce.

But, I did find an undesirable behavior. If you have a Force-Shielded creature, and its shield gets depleted partially, and then you cast Force Shield again, we'd like it to be the case that its shield gets replenished. Currently, that does not happen.

File: src/cpp/shared/combat/army.cpp (hunk 3 above)
Comment (Kert) [reply src/cpp/shared/combat/army.cpp]: Good catch. But yes it works, just checked

**Comment by Kert** [comment]
> In some of my early tests, I saw the mass spell not being cast on dwarves. However, in later testing, I was unable to reproduce.

Dwarves naturally have a 25% chance to block most spells. To exclude a beneficial spell importing army::SpellCastWorkChance will be required because there are a lot of corner cases there. It's actually not easy to import

> But, I did find an undesirable behavior. If you have a Force-Shielded creature, and its shield gets depleted partially, and then you cast Force Shield again, we'd like it to be the case that its shield gets replenished. Currently, that does not happen.

Fixed

File: src/cpp/shared/spell/spells.cpp (hunk 4 above)
Comment (Kert) [reply src/cpp/shared/spell/spells.cpp]: Fixed

**Comment by Kert** [comment]
Fixed the issue of dwarves having a chance to resist Force Shield by importing army::SpellCastWorkChance. I am not sure how to test all the spells after the import because something could be broken in the process. Original code had a bunch of goto-s

**Comment by jkoppel** [comment]
So, if there's a Phoenix with a 99 HP Force Shield on it, and you cast Force Shield, will it now have a 100 HP Force Shield?

If yes, then merge this code.

**Comment by Kert** [comment]
> So, if there's a Phoenix with a 99 HP Force Shield on it, and you cast Force Shield, will it now have a 100 HP Force Shield?
> 
> If yes, then merge this code.

It will be also casted even it has Full HP Force Shield. Should I change it?

**Comment by jkoppel** [comment]
Does this spell have a duration?

If yes, then merge. If no, then prevent the player from doing that, and then merge.

**Comment by Kert** [comment]
> Does this spell have a duration?
> 
> If yes, then merge. If no, then prevent the player from doing that, and then merge.

It works until it breaks, so yeah it has no duration. Disallowed casting the spell when hp is full

**Pull request merged**